File: apps/api/schemas.py
# apps/api/schemas.py
from __future__ import annotations
from typing import Dict, List, Optional, Literal
from pydantic import BaseModel, Field

Rarity = Literal["Common", "Magic", "Rare", "Epic", "Legendary", "Mythic"]

class OrbIn(BaseModel):
    type: str
    set: str
    rarity: str
    value: float
    level: int
    awakened: int = 0

class OrbOut(BaseModel):
    type: str
    set: str
    rarity: Rarity
    value: float
    level: int
    awakened: int = 0
    slot_index: Optional[int] = None  # allowed but optional

class SharedSlotProfileImpactOut(BaseModel):
    profile: str
    selected_score: float = 0.0
    solo_best_score: float = 0.0
    compromise_loss: float = 0.0
    selected_d_set: float = 0.0
    selected_d_orb: float = 0.0
    solo_best_d_set: float = 0.0
    solo_best_d_orb: float = 0.0
    cap_limited: bool = False

class SharedSlotAssignmentOut(BaseModel):
    category: str
    slot_index: int
    profiles: List[str] = Field(default_factory=list)
    is_uniform: bool = False
    orb: Optional[OrbOut] = None
    profile_orbs: Dict[str, Optional[OrbOut]] = Field(default_factory=dict)
    profile_impacts: List[SharedSlotProfileImpactOut] = Field(default_factory=list)

class SharedSummaryOut(BaseModel):
    requested_slots: int = 0
    filled_slots: int = 0
    is_partial: bool = False
    requested_positions: int = 0
    filled_positions: int = 0
    active_sets: Dict[str, int] = Field(default_factory=dict)
    totals_by_type: Dict[str, float] = Field(default_factory=dict)
    compromise_loss_total: float = 0.0
    compromise_loss_by_profile: Dict[str, float] = Field(default_factory=dict)
    cap_limited_slots_by_profile: Dict[str, int] = Field(default_factory=dict)
    slots: List[SharedSlotAssignmentOut] = Field(default_factory=list)

class ProfileRaw(BaseModel):
    name: str
    score: Optional[float] = None
    set_score: Optional[float] = None
    orb_score: Optional[float] = None
    requested_slots: int = 0
    filled_slots: int = 0
    is_partial: bool = False
    assignments: Dict[str, List[OrbOut]] = Field(default_factory=dict)

class RawPayload(BaseModel):
    combined_score: Optional[float] = None
    requested_slots: int = 0
    filled_slots: int = 0
    is_partial: bool = False
    shared_summary: Optional[SharedSummaryOut] = None
    profiles: List[ProfileRaw] = Field(default_factory=list)

class SummaryProfile(BaseModel):
    name: str
    score: Optional[float] = None
    set_score: Optional[float] = None
    orb_score: Optional[float] = None
    requested_slots: int = 0
    filled_slots: int = 0
    is_partial: bool = False

class SummaryPayload(BaseModel):
    combined_score: Optional[float] = None
    requested_slots: int = 0
    filled_slots: int = 0
    is_partial: bool = False
    shared_summary: Optional[SharedSummaryOut] = None
    profiles: List[SummaryProfile] = Field(default_factory=list)

class OptimizeProfileIn(BaseModel):
    name: str
    weight: float
    objective: Literal["sets-first", "types-first"]
    power: float
    epsilon: float
    set_priority: Dict[str, float]
    orb_weights: Dict[str, float]
    orb_level_weights: Dict[str, float]
    categories: Optional[Dict[str, Rarity]] = None
    slots: Optional[Dict[str, int]] = None  # direct override

class OptimizeRequest(BaseModel):
    orbs: List[OrbIn]
    profiles: List[OptimizeProfileIn]
    shareable_categories: Optional[List[str]] = None
    algorithm: Literal["greedy"] = "greedy"  # web app supports greedy only

class OptimizeResult(BaseModel):
    summary: SummaryPayload
    raw: RawPayload

class OptimizeResponse(BaseModel):
    ok: bool
    result: OptimizeResult

File: web/src/lib/schemas.ts
import { z } from "zod";
import {
  categoryRarityOptions,
  objectiveOptions,
  orbRarityOptions,
  type OptimizeRequest,
  type OptimizeResponse,
} from "@/lib/types";

const orbRaritySchema = z.enum(orbRarityOptions);
const categoryRaritySchema = z.enum(categoryRarityOptions);
const objectiveSchema = z.enum(objectiveOptions);

const numberOrZero = z.coerce.number().catch(0);
const integerOrZero = z.coerce.number().int().catch(0);

export const orbInSchema = z
  .object({
    type: z.string().trim().min(1),
    set: z.string().trim().min(1),
    rarity: orbRaritySchema.catch("Rare"),
    value: numberOrZero,
    level: integerOrZero,
    awakened: integerOrZero.default(0),
    slot_index: integerOrZero.optional(),
  })
  .transform((orb) => ({
    ...orb,
    value: Number.isFinite(orb.value) ? Math.max(0, orb.value) : 0,
    level: Math.max(0, orb.level),
    awakened: Math.max(0, orb.awakened),
  }));

export const optimizeProfileInSchema = z.object({
  name: z.string().trim().min(1),
  weight: z.coerce.number(),
  objective: objectiveSchema,
  power: z.coerce.number(),
  epsilon: z.coerce.number(),
  set_priority: z.record(z.string(), z.coerce.number()).default({}),
  orb_weights: z.record(z.string(), z.coerce.number()).default({}),
  orb_level_weights: z.record(z.string(), z.coerce.number()).default({}),
  categories: z.record(z.string(), categoryRaritySchema).optional(),
  slots: z.record(z.string(), z.coerce.number().int()).optional(),
});

export const optimizeRequestSchema = z.object({
  orbs: z.array(orbInSchema),
  profiles: z.array(optimizeProfileInSchema),
  shareable_categories: z.array(z.string()).optional(),
  algorithm: z.literal("greedy"),
});

const optimizeRawProfileSchema = z
  .object({
    name: z.string(),
    score: z.coerce.number().nullable().optional(),
    set_score: z.coerce.number().nullable().optional(),
    orb_score: z.coerce.number().nullable().optional(),
    used_slots: z.record(z.string(), z.coerce.number().int()).default({}),
    assignments: z.record(z.string(), z.array(orbInSchema).default([])).default({}),
  })
  .transform((profile) => ({
    ...profile,
    score: profile.score ?? null,
    set_score: profile.set_score ?? null,
    orb_score: profile.orb_score ?? null,
  }));

export const optimizeRawPayloadSchema = z
  .object({
    combined_score: z.coerce.number().nullable().optional(),
    profiles: z.array(optimizeRawProfileSchema).default([]),
  })
  .transform((payload) => ({
    ...payload,
    combined_score: payload.combined_score ?? null,
  }));

export const optimizeSummarySchema = z
  .object({
    combined_score: z.coerce.number().nullable().optional(),
    per_profile: z
      .array(
        z.object({
          name: z.string(),
          score: z.coerce.number().nullable().optional(),
          set_score: z.coerce.number().nullable().optional(),
          orb_score: z.coerce.number().nullable().optional(),
        })
      )
      .default([]),
  })
  .transform((summary) => ({
    combined_score: summary.combined_score ?? null,
    per_profile: summary.per_profile.map((profile) => ({
      ...profile,
      score: profile.score ?? null,
      set_score: profile.set_score ?? null,
      orb_score: profile.orb_score ?? null,
    })),
  }));

export const optimizeResponseSchema = z.object({
  ok: z.literal(true),
  result: z.object({
    summary: optimizeSummarySchema,
    raw: optimizeRawPayloadSchema,
  }),
});

export function parseOptimizeRequest(value: unknown): OptimizeRequest {
  return optimizeRequestSchema.parse(value);
}

export function parseOptimizeResponse(value: unknown): OptimizeResponse {
  return optimizeResponseSchema.parse(value);
}
